### PR TITLE
[BugFix] move _overflow_read_chunk into scan context

### LINKED ISSUE #5933: BE crash
### Steps to reproduce the behavior (Required)
1. CREATE TABLE `test_all_type_select` (
  `id_int` int(11) NOT NULL COMMENT "",
  `id_int1000` int(11) NOT NULL COMMENT "",
  `id_tinyint` tinyint(4) NOT NULL COMMENT "",
  `id_smallint` smallint(6) NOT NULL COMMENT "",
  `id_bigint` bigint(20) NOT NULL COMMENT "",
  `id_largeint` largeint(40) NOT NULL COMMENT "",
  `id_float` float NOT NULL COMMENT "",
  `id_double` double NOT NULL COMMENT "",
  `id_char` char(20) NOT NULL COMMENT "",
  `id_varchar` varchar(100) NOT NULL COMMENT "",
  `id_varchar_int` varchar(100) NOT NULL COMMENT "",
  `id_varchar_date` varchar(100) NOT NULL COMMENT "",
  `id_varchar_datetime` varchar(100) NOT NULL COMMENT "",
  `id_date` date NOT NULL COMMENT "",
  `id_datetime` datetime NOT NULL COMMENT "",
  `id_decimal` decimal128(20, 3) NOT NULL COMMENT "",
  `id_decimal10` decimal128(20, 9) NOT NULL COMMENT "",
  `id_boolean` boolean NOT NULL COMMENT "",
  `nid_int` int(11) NULL COMMENT "",
  `nid_int1000` int(11) NULL COMMENT "",
  `nid_tinyint` tinyint(4) NULL COMMENT "",
  `nid_smallint` smallint(6) NULL COMMENT "",
  `nid_bigint` bigint(20) NULL COMMENT "",
  `nid_largeint` largeint(40) NULL COMMENT "",
  `nid_float` float NULL COMMENT "",
  `nid_double` double NULL COMMENT "",
  `nid_char` char(20) NULL COMMENT "",
  `nid_varchar` varchar(100) NULL COMMENT "",
  `nid_varchar_int` varchar(100) NULL COMMENT "",
  `nid_varchar_date` varchar(100) NULL COMMENT "",
  `nid_varchar_datetime` varchar(100) NULL COMMENT "",
  `nid_date` date NULL COMMENT "",
  `nid_datetime` datetime NULL COMMENT "",
  `nid_decimal` decimal128(20, 3) NULL COMMENT "",
  `nid_decimal10` decimal128(20, 9) NULL COMMENT "",
  `nid_boolean` boolean NULL COMMENT ""
) ENGINE=OLAP
DUPLICATE KEY(`id_int`)
COMMENT "OLAP"
DISTRIBUTED BY HASH(`id_int`) BUCKETS 10
PROPERTIES (
"replication_num" = "1",
"in_memory" = "false",
"storage_format" = "DEFAULT",
"enable_persistent_index" = "false"
);
2. select count(*) from ( SELECT id_bigint, id_double, id_float, id_char, id_int, count(*) from test_all_type_select where id_smallint > 23 GROUP by id_bigint, id_double, id_float, id_char , id_int) AS a;

### Expected behavior (Required)
Query returns correctly

### Real behavior (Required)
```
*** Aborted at 1652237598 (unix time) try "date -d @1652237598" if you are using GNU date ***
PC: @          0x1c4aeb1 starrocks::ScalarColumnIterator::next_batch()
*** SIGSEGV (@0x0) received by PID 3121 (TID 0x7f949e1a0700) from PID 0; stack trace: ***
    @          0x3eded32 google::(anonymous namespace)::FailureSignalHandler()
    @     0x7f94cf7b3630 (unknown)
    @          0x1c4aeb1 starrocks::ScalarColumnIterator::next_batch()
    @          0x1aa7180 starrocks::vectorized::SegmentIterator::_read()
    @          0x1a9f665 starrocks::vectorized::SegmentIterator::_do_get_next()
    @          0x1aa2de1 starrocks::vectorized::SegmentIterator::do_get_next()
    @          0x1afe642 starrocks::vectorized::ProjectionIterator::do_get_next()
    @          0x1b3a5f4 starrocks::vectorized::UnionIterator::do_get_next()
    @          0x1f23afa starrocks::SegmentIteratorWrapper::do_get_next()
    @          0x1b3494b starrocks::vectorized::TimedChunkIterator::do_get_next()
    @          0x1b2d37e starrocks::vectorized::TabletReader::do_get_next()
    @          0x29077cc starrocks::pipeline::OlapChunkSource::_read_chunk_from_storage()
    @          0x2907e50 starrocks::pipeline::OlapChunkSource::buffer_next_batch_chunks_blocking()
    @          0x28074e5 _ZNSt17_Function_handlerIFvvEZN9starrocks8pipeline12ScanOperator18_trigger_next_scanEPNS1_12RuntimeStateEiEUlvE0_E9_M_invokeERKSt9_Any_data
    @          0x1f48f90 starrocks::PriorityThreadPool::work_thread()
    @          0x3e7a367 thread_proxy
    @     0x7f94cf7abea5 start_thread
    @     0x7f94cedc6b0d __clone
    @                0x0 (unknown)
```

### StarRocks version (Required)
 - You can get the StarRocks version by executing SQL `select current_version()`
main release 8b395a8

### LINKED ISSUE #5934: BE crash.
### Steps to reproduce the behavior (Required)
1. CREATE TABLE '...'
2. INSERT INTO '....'
3. SELECT  '....'
 tpcds 1g, query18.
```
select  i_item_id, ca_country, ca_state, ca_county, avg( cast(cs_quantity as decimal(12,2))) agg1, avg( cast(cs_list_price as decimal(12,2))) agg2, avg( cast(cs_coupon_amt as decimal(12,2))) agg3, avg( cast(cs_sales_price as decimal(12,2))) agg4, avg( cast(cs_net_profit as decimal(12,2))) agg5, avg( cast(c_birth_year as decimal(12,2))) agg6, avg( cast(cd1.cd_dep_count as decimal(12,2))) agg7 from catalog_sales, customer_demographics cd1, customer_demographics cd2, customer, customer_address, date_dim, item where cs_sold_date_sk = d_date_sk and cs_item_sk = i_item_sk and cs_bill_cdemo_sk = cd1.cd_demo_sk and cs_bill_customer_sk = c_customer_sk and cd1.cd_gender = 'F' and cd1.cd_education_status = 'Unknown' and c_current_cdemo_sk = cd2.cd_demo_sk and c_current_addr_sk = ca_address_sk and c_birth_month in (1, 6, 8, 9, 12, 2) and d_year = 1998 and ca_state in ('MS','IN','ND' ,'OK','NM','VA','MS') group by rollup (i_item_id, ca_country, ca_state, ca_county) order by ca_country, ca_state, ca_county, i_item_id limit 100;  
```
### Real behavior (Required)
```
start time: Wed May 11 10:59:20 CST 2022
*** Aborted at 1652237999 (unix time) try "date -d @1652237999" if you are using GNU date ***
PC: @          0x1c4aeb1 starrocks::ScalarColumnIterator::next_batch()
*** SIGSEGV (@0x0) received by PID 21469 (TID 0x7f691839a700) from PID 0; stack trace: ***
    @          0x3eded32 google::(anonymous namespace)::FailureSignalHandler()
    @     0x7f694cdeb630 (unknown)
    @          0x1c4aeb1 starrocks::ScalarColumnIterator::next_batch()
    @          0x1aa7180 starrocks::vectorized::SegmentIterator::_read()
    @          0x1a9f665 starrocks::vectorized::SegmentIterator::_do_get_next()
    @          0x1aa2de1 starrocks::vectorized::SegmentIterator::do_get_next()
    @          0x1afe642 starrocks::vectorized::ProjectionIterator::do_get_next()
    @          0x1f23afa starrocks::SegmentIteratorWrapper::do_get_next()
    @          0x1b3494b starrocks::vectorized::TimedChunkIterator::do_get_next()
    @          0x1b2d37e starrocks::vectorized::TabletReader::do_get_next()
    @          0x29077cc starrocks::pipeline::OlapChunkSource::_read_chunk_from_storage()
    @          0x2907e50 starrocks::pipeline::OlapChunkSource::buffer_next_batch_chunks_blocking()
    @          0x28074e5 _ZNSt17_Function_handlerIFvvEZN9starrocks8pipeline12ScanOperator18_trigger_next_scanEPNS1_12RuntimeStateEiEUlvE0_E9_M_invokeERKSt9_Any_data
    @          0x1f48f90 starrocks::PriorityThreadPool::work_thread()
    @          0x3e7a367 thread_proxy
    @     0x7f694cde3ea5 start_thread
    @     0x7f694c3fe8dd __clone
    @                0x0 (unknown)
```
### StarRocks version (Required)
 - You can get the StarRocks version by executing SQL `select current_version()`
 - MAIN-RELEASE 8b395a8

### PR DESCRIPTION
**Pull request by xiaoyong-z**
Signed-off-by: xyz <a997647204@gmail.com>

## What type of PR is this：
- [x] bug
- [ ] feature
- [ ] enhancement
- [ ] others

## Which issues of this PR fixes ：

Fixes #5934 
Fixes #5933 

## Problem Summary(Required) ：

This pr fix the bug introduced by #5701

Under late materialization scenario, context 0 and context 1 have different read schema and column iterators. The overflow rows in _overflow_read_chunk by context0 will be used by context1 and leads to a crash.

This pr moves the _overflow_read_chunk into scan context, each context has its own _overflow_read_chunk.

### CODE REVIEW

File: be/src/storage/rowset/segment_iterator.cpp [hunk 1]
```diff
@@ -675,12 +674,14 @@ Status SegmentIterator::_do_get_next(Chunk* result, vector<rowid_t>* rowid) {
     // If _overflow_read_chunk contains some rows, it means that in previous round of `_do_get_next`,
     // _read_chunk reads more rows than its capacity, the overflow rows in saved in _overflow_read_chunk.
     // Here we reload those overflow rows to current context's _read_chunk.
-    if (_overflow_read_chunk != nullptr && !_overflow_read_chunk->is_empty()) {
-        _context->_read_chunk->swap_chunk(*_overflow_read_chunk);
+    if (_context->_overflow_read_chunk != nullptr && !_context->_overflow_read_chunk->is_empty()) {
```
Comment (kangkaisen) [inline be/src/storage/rowset/segment_iterator.cpp:677]:
```
    // remove (logical) deleted rows.
    if (chunk_size > 0 && chunk->delete_state() != DEL_NOT_SATISFIED && !_opts.delete_predicates.empty()) {
```
does _overflow_read_chunk need handle delete state?

**Comment by kangkaisen** [comment]
This PR is running daily test.

File: be/src/storage/rowset/segment_iterator.cpp [hunk 2]
```diff
@@ -166,6 +166,10 @@ class SegmentIterator final : public ChunkIterator {
 
         // not all dict encode
         bool _has_force_dict_encode{false};
+
+        // save overflow rows in read_chunk
+        std::shared_ptr<Chunk> _overflow_read_chunk;
```
Comment (trueeyu) [inline be/src/storage/rowset/segment_iterator.cpp:171]: _overflow_read_chunk seems that it has never been used, it is all used context->_overflow_read_chunk

Comment (trueeyu) [inline be/src/storage/rowset/segment_iterator.cpp:171]: the `_overflow_read_chunk` in SegmentIterator have never used again?

**Review by trueeyu — DISMISSED** [review]

File: be/src/storage/rowset/segment_iterator.cpp (hunk 1 above)
Comment (xiaoyong-z) [reply be/src/storage/rowset/segment_iterator.cpp:677]: i think it needs to

Comment (xiaoyong-z) [reply be/src/storage/rowset/segment_iterator.cpp:677]: current code haven't dealt with this yet

**Review by chaoyli — DISMISSED** [review]

**Comment by stdpain** [comment]
could you add a flag to control overflow_read ? I'm not sure the effect for large TYPE (ARRAY/BITMAP/HLL/JSON)

**Review by chaoyli — APPROVED** [review]

**Review by kangkaisen — APPROVED** [review]

**Pull request merged**